src/ext/php7/curl_handlers.c

```diff
@@ -0,0 +1,612 @@
+#include "curl_handlers.h"
+
+#include <Zend/zend_interfaces.h>
+#include <php.h>
+
+#include "compat_string.h"
+#include "configuration.h"
+#include "engine_hooks.h"  // for ddtrace_backup_error_handling
+
+#ifndef ZVAL_COPY_DEREF
+#define ZVAL_COPY_DEREF(z, v)                     \
+    do {                                          \
+        zval *_z3 = (v);                          \
+        if (Z_OPT_REFCOUNTED_P(_z3)) {            \
+            if (UNEXPECTED(Z_OPT_ISREF_P(_z3))) { \
+                _z3 = Z_REFVAL_P(_z3);            \
+                if (Z_OPT_REFCOUNTED_P(_z3)) {    \
+                    Z_ADDREF_P(_z3);              \
+                }                                 \
+            } else {                              \
+                Z_ADDREF_P(_z3);                  \
+            }                                     \
+        }                                         \
+        ZVAL_COPY_VALUE(z, _z3);                  \
+    } while (0)
+#endif
+
+bool _dd_ext_curl_loaded = false;  // True global -- do not modify after startup
+
+/* "le_curl" is ext/curl's resource type.
+ * "le_curl" is what php_curl.h names this variable
+ */
+ZEND_TLS int le_curl = 0;
+
+/* Cache things we tend to use a few times */
+ZEND_TLS zval _dd_curl_httpheaders = {.u1.type_info = IS_UNDEF};
+ZEND_TLS zval _dd_format_curl_http_headers = {.u1.type_info = IS_UNDEF};
+ZEND_TLS zend_class_entry *_dd_ArrayKVStore_ce = NULL;
+ZEND_TLS zend_class_entry *_dd_GlobalTracer_ce = NULL;
+ZEND_TLS zend_class_entry *_dd_SpanContext_ce = NULL;
+ZEND_TLS zend_function *_dd_ArrayKVStore_putForResource_fe = NULL;
+ZEND_TLS zend_function *_dd_ArrayKVStore_getForResource_fe = NULL;
+ZEND_TLS zend_function *_dd_ArrayKVStore_deleteResource_fe = NULL;
+ZEND_TLS zend_function *_dd_GlobalTracer_get_fe = NULL;
+ZEND_TLS zend_function *_dd_GlobalTracer_inject_fe = NULL;
+ZEND_TLS zend_function *_dd_SpanContext_ctor = NULL;
+ZEND_TLS bool _dd_curl_integration_loaded = false;
+
+// Do not pass things like "parent", "self", "static" -- fully qualified names only!
+static zend_class_entry *_dd_lookup_ce(const char *str, size_t len, bool autoload) {
```

Comment:
May I recommend removing `bool autoload` from the function signature? I think we should always call `zend_lookup_class_ex()` with `ZEND_FETCH_CLASS_NO_AUTOLOAD`. The reason is that later on we might add something that accidentally triggers the autoloader with this API which could have adverse effects. I think if we need autoloading in the future it should have a different function signature and be named in such a way that makes it explicit that we are invoking the autoloader on purpose. Something like `_dd_dangerously_lookup_ce_with_autoloading_enabled()` or something. 😆

Reply:
I'll think about it. In any case, we only use the autoload=0 in this PR, so it's untested API surface to actually autoload.

Reply:
I removed the autoload bit for now.